src/ai/backend/agent/resources.py

```diff
@@ -523,6 +523,53 @@ async def _wrap_future(fut: asyncio.Future) -> None:
     return slot_allocs
 
 
+async def scan_gpu_alloc_map(
+    kernel_ids: Sequence[KernelId],
+    scratch_root: Path,
+) -> dict[DeviceId, Decimal]:
+    """
+    Fetch the current allocated amounts for fractional gpu from
+    ``/home/config/resource.txt`` files in the kernel containers managed by this agent.
+    """
+    loop = asyncio.get_running_loop()
+    gpu_alloc_map: dict[DeviceId, Decimal] = defaultdict(lambda: Decimal(0))
+
+    def _read_kernel_resource_spec(path: Path) -> None:
+        nonlocal gpu_alloc_map
+        try:
+            resource_spec = KernelResourceSpec.read_from_string(path.read_text())
+
+            if cuda := resource_spec.allocations.get(DeviceName("cuda")):
+                if cuda_shares := cuda.get(SlotName("cuda.shares")):
+                    for device_id, shares in cuda_shares.items():
+                        gpu_alloc_map[device_id] += Decimal(shares)
+
+                if cuda_device := cuda.get(SlotName("cuda.device")):
+                    for device_id, device in cuda_device.items():
+                        gpu_alloc_map[device_id] += Decimal(device)
```

Comment:
I think it would be problematic to update the same dict when it will be executed in a different coroutine.
It would be better to generate `alloc_map` in `_read_kernel_resource_spec`, act asynchronously throughout, and then combine them.